CMakeLists.txt

```diff
@@ -93,37 +38,78 @@ if(NOT EXECUTABLE_OUTPUT_PATH)
-message(STATUS "CMAKE_SOURCE_DIR: " ${CMAKE_SOURCE_DIR})
-message(STATUS "CMAKE_TEST_DIR: " ${CMAKE_TEST_DIR})
-message(STATUS "RUN TIME " ${CMAKE_RUNTIME_OUTPUT_DIRECTORY})
-message(STATUS "EXECUTABE" ${EXECUTABLE_OUTPUT_PATH})
-# ========================== build platform ==========================
-message(STATUS "")
-message(STATUS "Platform: ")
-message(STATUS "    Host: " ${CMAKE_HOST_SYSTEM_NAME} ${CMAKE_HOST_SYSTEM_VERSION} ${CMAKE_HOST_SYSTEM_PROCESSOR})
-if (CMAKE_CROSSCOMPILING)
-    message(STATUS "    Target: " ${CMAKE_SYSTEM_NAME} ${CMAKE_SYSTEM_VERSION} ${CMAKE_SYSTEM_PROCESSOR})
-endif ()
-message(STATUS "    CMake: " ${CMAKE_VERSION})
-message(STATUS "    CMake generator: " ${CMAKE_GENERATOR})
-message(STATUS "    CMake build tool: " ${CMAKE_BUILD_TOOL})
-if (MSVC)
-    message(STATUS "    MSVC: " ${MSVC_VERSION})
-#set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ZLIB_WINAPI)
-endif ()
-if (CMAKE_GENERATOR MATCHES Xcode)
-    message(STATUS "    Xcode: " ${XCODE_VERSION})
-endif ()
-if (NOT CMAKE_GENERATOR MATCHES "Xcode|Visual Studio")
-    message(STATUS "    Configuration: " ${CMAKE_BUILD_TYPE})
-endif ()
-message(STATUS "")
-
-find_package(Git QUIET) # no need for this msg. It comes from cmake.findgit()
+
+################################################################################
+# Optional features
+################################################################################
+set(BUILD_TESTING OFF CACHE BOOL "Build with integration tests")
+print_var(BUILD_TESTING)
+
+set(BUILD_CPPUNIT_TEST OFF CACHE BOOL "Build with unit tests")
+print_var(BUILD_CPPUNIT_TEST)
+
+set(BUILD_DOC OFF CACHE BOOL "Build doxygen documentation")
+print_var(BUILD_DOC)
+
+set(JPSFIRE OFF CACHE BOOL "Build with jpsfire support")
+print_var(JPSFIRE)
+
+set(AIROUTER OFF CACHE BOOL "Build with AI router support")
+print_var(AIROUTER)
+
+################################################################################
+# Compilation flags
+################################################################################
+# Note: Setting global compile flags via CMAKE_CXX_FLAGS has the drawback that
+#       generator expressions cannot be used. This leads to all kind of
+#       conditional adding of flags. It is generally preferable to use generator
+#       expresssions.
+#
+# WARNING: Do not break the lines, each option has to be on its own line or
+#          CMake will enclose multiple flags in '' which the compiler then
+#          treats as a single flag and does not understand.
+list(APPEND COMMON_COMPILE_OPTIONS
+    $<$<OR:$<CXX_COMPILER_ID:Clang>,$<CXX_COMPILER_ID:AppleClang>,$<CXX_COMPILER_ID:GNU>>:-Wall>
+    $<$<OR:$<CXX_COMPILER_ID:Clang>,$<CXX_COMPILER_ID:AppleClang>,$<CXX_COMPILER_ID:GNU>>:-Wextra>
+    $<$<OR:$<CXX_COMPILER_ID:Clang>,$<CXX_COMPILER_ID:AppleClang>,$<CXX_COMPILER_ID:GNU>>:-fdiagnostics-color=always>
+    $<$<CXX_COMPILER_ID:MSVC>:/W4>
+    $<$<CXX_COMPILER_ID:MSVC>:/WX>
```

Comment:
Works good on windows with VS2017.
When compile with VS2017, may be /WX- instead of /WX.
And I have two questions :  
Do we need to set the runtime library ? (/MT /MD /MTd /MDd)
How to set the compile flags for third-party?

Reply:
@xuqiancheng you can try without `/MT /MD /MTd /MDd` and see what happens

Reply:
@xuqiancheng good point about the compile flags for third-party, ill make a patch to pass them on

Reply:
With the last two commits added there is still one thing left I would like to address: 

The macports/brew handling: For brew this seems unnecessary since libs installed with brew are found unless the conflict with system libs. In that case brew gives clear instructions on what to do.

For macports users should pass -DCMAKE_PREFIX_PATH=<mymacports-lib-path> when invoking macports. See: https://cmake.org/cmake/help/v3.15/variable/CMAKE_PREFIX_PATH.html#variable:CMAKE_PREFIX_PATH

This is a little less convenient but gives uses more control over their library paths. 
```cmake
#http://stackoverflow.com/questions/1487752/how-do-i-instruct-cmake-to-look-for-libraries-installed-by-macports
if (APPLE)
    # Detect if the "port" command is valid on this system; if so, return full path
    execute_process(COMMAND which port RESULT_VARIABLE DETECT_MACPORTS OUTPUT_VARIABLE MACPORTS_PREFIX ERROR_QUIET OUTPUT_STRIP_TRAILING_WHITESPACE)
    if (${DETECT_MACPORTS} EQUAL 0)
        # "/opt/local/bin/port" doesn't have libs, so we get the parent directory
        get_filename_component(MACPORTS_PREFIX ${MACPORTS_PREFIX} DIRECTORY)
        # "/opt/local/bin" doesn't have libs, so we get the parent directory
        get_filename_component(MACPORTS_PREFIX ${MACPORTS_PREFIX} DIRECTORY)
        # "/opt/local" is where MacPorts lives, add `/lib` suffix and link
        link_directories(${MACPORTS_PREFIX}/lib)
        message(STATUS "Macports detected: ${MACPORTS_PREFIX}/lib")
        # SET(CMAKE_SYSTEM_NAME Darwin)
        # # Add MacPorts
        # INCLUDE_DIRECTORIES(/opt/local/include)

        # LINK_DIRECTORIES(/opt/local/lib)

    else ()
        # Recommendation, also add a "brew --prefix" custom command to detect a homebrew build environment
        execute_process(COMMAND brew --prefix RESULT_VARIABLE DETECT_BREW OUTPUT_VARIABLE BREW_PREFIX ERROR_QUIET OUTPUT_STRIP_TRAILING_WHITESPACE)
        if (${DETECT_BREW} EQUAL 0)
            link_directories(${BREW_PREFIX}/lib)
            message(STATUS "Brew detected: ${BREW_PREFIX}")
        endif ()
    endif ()
endif (APPLE)
```

Reply:
Eh to be clear I am proposing to just remove this section :)

Reply:
> @xuqiancheng you can try without `/MT /MD /MTd /MDd` and see what happens

I have tested it without setting the runtime library, and it works on windows.

Reply:
> @xuqiancheng good point about the compile flags for third-party, ~ill make a patch to pass them on~ On second thought, what do you want to pass on? The stricter warning flags do more harm when applied to third parties. You will see warnings but rarely can do something about them.

Many thanks! It works nice on windows even without setting /MT /MD /MTd /MDd.
I propose this suggestion is because I have met some link errors when I compile jpscore with vs2017, which are some mismatch between the third-party, jpcore and core. And the reason for these mismatch is different runtime libraries. So I think it would be better to specific the runtime libraries instead of default to avoid these mismatch, although now  it works nice with default setting.

Reply:
> Eh to be clear I am proposing to just remove this section :)

@ozaq: Sorry I missed this. You removed the brew/port section, but yet we still have some port-specific code there. See #445 .

It's just needed for the OpenMP-case, right?